### PR TITLE
Return signatures of Autoloader's loaders should be void

### PR DESCRIPTION
**Pull request by paulbalandan**
**Description**
Ref: https://www.php.net/manual/en/function.spl-autoload-register.php

As a result of this change, this may be a potential BC break. I added this to the changelog. Also marked `@internal` those 2 callbacks.

**Checklist:**
- [x] Securely signed commits
- [ ] Component(s) with PHPDoc blocks, only if necessary or adds value
- [x] Unit testing, with >80% coverage
- [ ] User guide updated
- [x] Conforms to style guide

### CODE REVIEW

File: system/Autoloader/Autoloader.php [hunk 1]
```diff
@@ -238,33 +238,34 @@ public function removeNamespace(string $namespace)
...
-        return false;
     }
 
     /**
      * Loads the class file for a given class name.
      *
+     * @internal For `spl_autoload_register` use.
+     *
      * @param string $class The fully qualified class name.
      *
```
Comment (mostafakhudair) [inline system/Autoloader/Autoloader.php:260]:
```suggestion
```

File: system/Autoloader/Autoloader.php [hunk 2]
```diff
@@ -238,33 +238,34 @@ public function removeNamespace(string $namespace)
     /**
      * Load a class using available class mapping.
      *
-     * @return false|string
+     * @internal For `spl_autoload_register` use.
+     *
+     * @return void
      */
     public function loadClassmap(string $class)
```
Comment (mostafakhudair) [inline system/Autoloader/Autoloader.php:245]:
```suggestion
     */
    public function loadClassmap(string $class): void
```

File: system/Autoloader/Autoloader.php [hunk 3]
```diff
@@ -238,33 +238,34 @@ public function removeNamespace(string $namespace)
...
      *
+     * @internal For `spl_autoload_register` use.
+     *
      * @param string $class The fully qualified class name.
      *
-     * @return false|string The mapped file on success, or boolean false
-     *                      on failure.
+     * @return void
      */
     public function loadClass(string $class)
```
Comment (mostafakhudair) [inline system/Autoloader/Autoloader.php:263]:
```suggestion
     */
    public function loadClass(string $class): void
```

**Review by mostafakhudair — APPROVED** [review]

**Comment by kenjis** [comment]
@paulbalandan I don't see the benefit. I don't think this is a bug.
Returning a value does not cause errors. I don't see that the callback return type must be void in PHP manual.
If you really want to change it, it is better to send to `4.4`.

I think adding `@internal` is okay.

File: system/Autoloader/Autoloader.php (hunk 1 above)
Comment (kenjis) [reply system/Autoloader/Autoloader.php:260]: Our coding rule adds the empty line.

**Comment by paulbalandan** [comment]
The signature is in the Parameters section:
> `callback`
> The autoload function being registered. If `null`, then the default implementation of [spl_autoload()](https://www.php.net/manual/en/function.spl-autoload.php) will be registered.
>
> `callback(string $class): void
`
> The `class` will not contain the leading backslash of a fully-qualified identifier.

Okay, I'll send to `4.4`

**Comment by kenjis** [comment]
@paulbalandan Okay, the callback return type should be `void`.

File: tests/system/Autoloader/AutoloaderTest.php [hunk 4]
```diff
@@ -50,13 +56,15 @@ protected function setUp(): void
 
         $this->loader = new Autoloader();
         $this->loader->initialize($config, $modules)->register();
+
+        $this->classLoader = $this->getPrivateMethodInvoker($this->loader, 'loadInNamespace');
```
Comment (kenjis) [inline tests/system/Autoloader/AutoloaderTest.php:60]: Why do you need to test protected `loadInNamespace()`?

Comment (kenjis) [reply tests/system/Autoloader/AutoloaderTest.php:60]: Out of the scope of this PR, but would the following code be necessary?
It seems not, and the `loadInNamespace()` method doesn't seem necessary either.

https://github.com/codeigniter4/CodeIgniter4/blob/31b79b163528c7ee0f27ebad52f13c890186a994/system/Autoloader/Autoloader.php#L263-L264

Comment (kenjis) [reply tests/system/Autoloader/AutoloaderTest.php:60]: I think we don't need this test either:
https://github.com/codeigniter4/CodeIgniter4/blob/31b79b163528c7ee0f27ebad52f13c890186a994/tests/system/Autoloader/AutoloaderTest.php#L149-L156

PHP autoloader never passes the value `'\App\Controllers\Home.php'`.

Comment (paulbalandan) [reply tests/system/Autoloader/AutoloaderTest.php:60]: > Out of the scope of this PR, but would the following code be necessary?
> It seems not, and the `loadInNamespace()` method doesn't seem necessary either.

Those are necessary as the callback for `spl_autoload_register` does not expect leading `\`. For the trimming of `.php` part I think not. I think it's a misuse that a classname ends with .php

Comment (paulbalandan) [reply tests/system/Autoloader/AutoloaderTest.php:60]: > I think we don't need this test either:
> 
> https://github.com/codeigniter4/CodeIgniter4/blob/31b79b163528c7ee0f27ebad52f13c890186a994/tests/system/Autoloader/AutoloaderTest.php#L149-L156
> 
> PHP autoloader never passes the value `'\App\Controllers\Home.php'`.

This test is to checks if the value passed to `loadInNamespace` is valid (i.e., a class with no leading `\` and no `.php` part). So basically this is what the previous test is.

Comment (kenjis) [reply tests/system/Autoloader/AutoloaderTest.php:60]: What do you mean?
Is there a way to pass the classname with leading `\`?
https://3v4l.org/THIOb

Comment (paulbalandan) [reply tests/system/Autoloader/AutoloaderTest.php:60]: Okay, I think I misunderstood the manual. In that case, the removal of leading `\` seems unnecessary. How about the removal of `.php` extension ?

Comment (kenjis) [reply tests/system/Autoloader/AutoloaderTest.php:60]: We don't need the removal of `.php` extension. 
It is not a valid classname, so the method never receive such string.

**Review by kenjis — APPROVED** [review]

**Pull request merged**